frappe/model/rename_doc.py

```diff
@@ -47,11 +63,33 @@ def update_document_title(
 	name_updated = updated_name and (updated_name != doc.name)
 
 	if name_updated:
-		docname = rename_doc(doctype=doctype, old=docname, new=updated_name, merge=merge)
+		if enqueue and not is_scheduler_inactive():
+			current_name = doc.name
+
+			# before_name hook may have DocType specific validations or transformations
+			transformed_name = doc.run_method("before_rename", current_name, updated_name, merge)
```

Comment:
potential of non-atomic updates here:

Example: https://github.com/frappe/erpnext/blob/707f624aabb6721d6a1817f75dbc1c17f8c5cf81/erpnext/stock/doctype/item/item.py#L395-L403

Reply:
Maybe we should think of UX of validations later and ensure atomicity first 🥲 

Or new hook `validate_hook` that just does validation and no DB writes.

Reply:
😭